File: scripts/cancel-neural-inference.mjs
import { promises as fs } from 'node:fs';
import path from 'node:path';
import { createRequire } from 'node:module';
import { createHash } from 'node:crypto';
import assert from 'node:assert/strict';
const [consumer,source,runtime,workspace,marker,output]=process.argv.slice(2);
const require=createRequire(path.resolve(consumer,'package.json'));
const {Digitizer}=require('ecg-waveparse');
const digitizer=new Digitizer({workspaceDir:workspace,runtimeDir:runtime});
const controller=new AbortController();
let observed,requestedAt,runId;
const sourceHash=createHash('sha256').update(await fs.readFile(source)).digest('hex');
const poll=setInterval(async()=>{
  if (observed) return;
  try { observed=JSON.parse(await fs.readFile(marker,'utf8')); }
  catch(error) { if(error.code==='ENOENT'||error instanceof SyntaxError) return; throw error; }
  requestedAt=performance.now();controller.abort();
},50);
let error;
try { await digitizer.digitize(source,{device:'cpu',timeoutMs:600000,signal:controller.signal,onProgress:event=>{runId=event.runId;}}); }
catch(caught) { error=caught; }
finally {clearInterval(poll);}
const latency=requestedAt===undefined?null:performance.now()-requestedAt;
assert(observed,'The experiment never observed actual convolution work.');
assert.equal(error?.code,'cancelled');assert(latency<30000);
const run=await digitizer.getRun(runId);
assert.equal(run.processing.state,'failed');assert.equal(run.status,'failed');
assert.equal(run.sourceIdentity.sha256,sourceHash);assert(!run.assets.canonicalCsv);
const evidence=await digitizer.getEvidence(runId);assert.equal(evidence.sourceInspectionAvailable,true);
let candidateProcessAlive=true;
try {process.kill(observed.pid,0);}catch(caught){if(caught.code==='ESRCH')candidateProcessAlive=false;else throw caught;}
assert.equal(candidateProcessAlive,false,'The cancelled neural process remains alive.');
await fs.writeFile(output,JSON.stringify({version:1,interface:'javascript',clinicalValidationUse:false,sourceSha256:sourceHash,
  boundary:observed.boundary,cancellationLatencyMs:latency,runId,status:run.status,processingState:run.processing.state,
  sourceInspectionAvailable:true,quantitativeOutputPresent:false,candidateProcessAlive,instrumentation:'One-shot profile hook; immutable package and model files unchanged.'},null,2)+'\n',{flag:'wx'});

File: scripts/cancel_neural_inference.py
"""Cancel the installed async Python client after observed neural computation."""
import asyncio
import hashlib
import json
import os
import sys
import time
from pathlib import Path
from ecg_waveparse import Digitizer


async def main():
    source,runtime,workspace,marker,output=sys.argv[1:]
    digitizer=Digitizer(workspace_dir=workspace,runtime_dir=runtime)
    source_hash=hashlib.sha256(Path(source).read_bytes()).hexdigest()
    task=asyncio.create_task(digitizer.digitize_async(source,device='cpu',timeout_ms=600000))
    observed=None
    while not task.done():
        try:observed=json.loads(Path(marker).read_text());break
        except (FileNotFoundError,json.JSONDecodeError):await asyncio.sleep(.05)
    assert observed,'The experiment never observed actual convolution work.'
    start=time.monotonic();task.cancel()
    try:await task;raise AssertionError('Cancellation was not propagated.')
    except asyncio.CancelledError:pass
    latency=(time.monotonic()-start)*1000
    assert latency<30000
    records=list((Path(workspace)/'storage/runs').glob('run_*/metadata.json'))
    assert len(records)==1
    run_id=json.loads(records[0].read_text())['id']
    run=await digitizer.get_run_async(run_id)
    assert run['processing']['state']==run['status']=='failed'
    assert run['sourceIdentity']['sha256']==source_hash and 'canonicalCsv' not in run['assets']
    evidence=await digitizer.get_evidence_async(run_id);assert evidence['sourceInspectionAvailable'] is True
    alive=True
    try:os.kill(observed['pid'],0)
    except ProcessLookupError:alive=False
    assert not alive,'The cancelled neural process remains alive.'
    result={'version':1,'interface':'python','clinicalValidationUse':False,'sourceSha256':source_hash,
            'boundary':observed['boundary'],'cancellationLatencyMs':latency,'runId':run_id,'status':run['status'],
            'processingState':run['processing']['state'],'sourceInspectionAvailable':True,'quantitativeOutputPresent':False,
            'candidateProcessAlive':alive,'instrumentation':'One-shot profile hook; immutable package and model files unchanged.'}
    with Path(output).open('x') as handle:json.dump(result,handle,indent=2)


if __name__=='__main__':asyncio.run(main())
